[Console] Fix autocomplete_entities API crash when response size is too big

mibragimov opened the pull request:
Fixes https://github.com/elastic/kibana/issues/144310

### Summary

This PR addresses the issue of the Kibana instance restarting when the response size is too big for the `autocomplete_entities` API. This happens when a cluster has a large number of mappings and we try to retrieve them all on the server side with `esClient.asInternalUser.indices.getMapping()`. esClient does not handle large responses well and throws an error that causes the Kibana instance to restart. As node's max [string length](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#description) is 2^28-1 (~512MB) if the response size is over 512MB, it will throw an error with the message "Invalid string length".

The fix is to use the raw http request to retrieve the mappings instead of esClient and check the response size before sending it to the client. If the response size is too big, we will return an empty object and log the error in the server logs.

#### Proposed changes

- Remove ES JS client requests and use native Node.js HTTP client instead
- Limit the response size to 10MB for the `autocomplete_entities` API

#### How to test this PR locally
To test this out, you will need to connect Kibana to a remote cluster with a large number of mappings. We created a patch file that you can apply to your local Kibana instance to test this PR. Since the patch file contains credentials, we can't share it publicly. Please reach out to me if you would like to test this PR. I will share the patch file and the instructions to apply it.


<!--ONMERGE {"backportTargets":["8.5"]} ONMERGE-->

alexwizp commented on src/plugins/console/server/routes/api/console/autocomplete_entities/register_get_route.ts:
```diff
@@ -70,16 +71,26 @@ export function registerGetRoute({ router, lib: { handleEsError } }: RouteDepend
 
-        const esClient = (await ctx.core).elasticsearch.client;
         const mappings = await getMappings(esClient, settings);
+        const maxMappingSizeExceeded = Object.keys(mappings).length > 10 * 1024 * 1024;
```
@cjcenizal  @sabarasaba I'm not sure that we should proceed with that changes. I see we are still overloading our ES instance, can we just limit the size of the data returned with displaying the warning message for that case?

Probably we can specify the next option with setting it to e.g. 1000
`maxResponseSize?: number;`

alexwizp commented on src/plugins/console/public/services/autocomplete.ts:
```diff
@@ -20,6 +21,8 @@ import {
 } from '../lib/autocomplete_entities';
 import { DevToolsSettings, Settings } from './settings';
 
+const AUTOCOMPLETE_TIMEOUT = 30000;
```
I vote we have a limit on the number of records returned instead of a timeout. In addition, if we still continue with a timeout, it must be on the server side so that we have the opportunity to abort requests that are too long

sabarasaba reviewed (commented):
Going through the changes and at first glance looks fine, I was wondering though how can I test this locally to make sure it works? 🤔

mibragimov commented:
> Going through the changes and at first glance looks fine, I was wondering though how can I test this locally to make sure it works? 🤔

Right. One way of testing this I found was to throttle the network in the Network tab of the Dev Tools. I left a note in the PR description about how to test this locally. Another way is to create a large number of indices, e.g. `PUT /test-1`, `PUT /test-2`, ..., `PUT /test-1000` and then send a request to Elasticsearch.

mibragimov commented:
@sabarasaba  Added a `ci:cloud-deploy` label to this PR to trigger a cloud deployment for testing. I am not sure but cloud might have more resources and it might be easier to reproduce the issue there.

alexwizp commented on src/plugins/console/public/services/autocomplete.ts:
```diff
@@ -28,53 +54,165 @@ export class AutocompleteInfo {
+    );
+
+    if (mappings) {
+      const maxMappingSize = Object.keys(mappings).length > 10 * 1024 * 1024;
```
Could you please explain? Looks like we already loaded all `mappings` here. 
It seems strange to me to upload 10MB of data to the client, and then show an error message and throw all this data away

mibragimov replied:
Addressed with https://github.com/elastic/kibana/pull/140569/commits/db281e37bf4177fce39e39b00d2502052a3e1fe9 as we discussed offline

alexwizp commented on src/plugins/console/public/services/autocomplete.ts:
```diff
@@ -28,53 +54,165 @@ export class AutocompleteInfo {
+  }
+
+  private async retrieveTemplates(settingsToRetrieve: SettingsToRetrieve) {
+    const legacyTemplates = await this.retrieveSettings<IndicesGetTemplateResponse>(
```
is it possible to run legacyTemplates, indexTemplates, componentTemplates in parallel?

mibragimov commented:
@sabarasaba @alexwizp  Added a few changes to protect against large mapping responses, to test this: 
1.  Set `MAX_MAPPING_RESPONSE_SIZE` e.g. to 1 in https://github.com/elastic/kibana/blob/main/src/plugins/console/server/lib/proxy_request.ts
2. Refresh the page to retrieve autocomplete suggestions. Ensure fields suggestions are enabled in Settings.

alexwizp commented on src/plugins/console/public/services/autocomplete.ts:
```diff
@@ -28,53 +54,174 @@ export class AutocompleteInfo {
+  }
+
+  private async retrieveTemplates(settingsToRetrieve: SettingsToRetrieve) {
+    const [legacyTemplates, indexTemplates, componentTemplates] = await Promise.all([
```
can it be handled independently?

```
this.retrieveSettings<IndicesGetTemplateResponse>(
  ENTITIES.LEGACY_TEMPLATES, settingsToRetrieve
).then(() => {
  if (legacyTemplates) {
    this.legacyTemplate.loadTemplates(legacyTemplates);
  }
}) 

```
and so on

I can explain in case of an error of some promise, the current code will not process successful requests, you can probably use allSettled, but I think this needs to be fixed somehow

alexwizp commented on src/plugins/console/server/lib/proxy_request.ts:
```diff
@@ -77,8 +81,39 @@ export const proxyRequest = ({
-    resolve(res);
+    // Check if the request is to _mapping endpoint and if so, limit the response to 10MB. This is to
+    // protect against large mapping responses that can cause the browser to hang.
+    if (isMappingEndpoint(pathname)) {
```
Please see documentation 

<img width="717" alt="image" src="https://user-images.githubusercontent.com/20072247/194523246-59b93648-0b06-4b9d-89a9-ec509ab41987.png">

Your code would make sense if callback works with partial data, in the current implementation it looks strange

@elastic/kibana-core do we have a legal way to limit request by payload size?

pgayvallet replied:
Sorry, I gonna need slightly more context here. What kind of request are we talking about (or are we talking about response here) ?

Requests to our endpoints? For payload-capable requests, the `options.body.maxBytes` (RouteConfigOptionsBody.maxBytes) can be used.

Requests performed from Kibana toward ES? No AFAIK we don't currently have an out-of-the-box way of doing it.

mibragimov replied:
Thanks for your reply @pgayvallet! We are talking about the response from es here. The problem we are trying to solve is that sometimes the response is too large and the browser hangs. This tends to happen for users with a large number of mappings. We are trying to find a way to limit the response size. What do you think about this  [approach](https://github.com/elastic/kibana/pull/140569/commits/3b75822ea4a569cf5578119d4da9564e24c1fd03#diff-dc97f758bd78991f73f2f928fc59d00923790286e12766dd7c91d148ec40cfeeR50) I'm not sure if there is a better way to do this.

pgayvallet replied:
First question: I know the console proxy requests are not using the ES library client, but are you forced to use an low level`http` client here instead of using the ES client exposed from Core given the ES endpoint you're reaching it not dynamic? If you were using the ES client lib to do your requests, you could use the `TransportRequestOptions.maxResponseSize` option to do what you want.


Now if you do need to use the same approach as other console ES calls, your approach looks fine. One question though: if the issue is about sending the response back to the client, do you really need to stream the raw response from ES to consume the size on each data emissions? Can't you just wait for the whole response and then look at it size to decide if you should be throwing or forwarding the response? It's less optimized for sure, but the implementation would be simpler.

mibragimov replied:
Only reason we're using the low-level http client is that the ES client doesn't seem to support limiting the response size. I've tried using the `maxResponseSize` option, but it doesn't seem to work. I've also tried using the `maxCompressedResponseSize` option, but that doesn't seem to work either. I could not find any use cases of these options in the codebase, except for the tests. I'm not sure if it's a bug or if I'm using it incorrectly.
                 
It makes sense to wait for the whole response and then look at its size. But would that be a problem if the response is very large? I'm not sure if that would cause a memory issue. What do you think? @pgayvallet @sabarasaba @alexwizp

mibragimov commented:
Closing this PR as the issue was fixed by https://github.com/elastic/kibana/pull/143428

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

mibragimov commented:
@elasticmachine merge upstream

alisonelizabeth commented:
@elasticmachine merge upstream

alisonelizabeth commented on src/plugins/console/server/routes/api/console/autocomplete_entities/index.ts:
```diff
@@ -6,4 +6,194 @@
+  return {};
+};
+
+const getEntity = (path: string, config: Config) => {
```
Can we add a code comment here to explain why we are using the raw http request to retrieve the entities instead of esClient?

alisonelizabeth commented on src/plugins/console/server/routes/api/console/autocomplete_entities/index.ts:
```diff
@@ -6,4 +6,194 @@
+
+          // If the request failed, log the error and return an empty object
+          if (result.reason instanceof Error) {
+            deps.log.error(result.reason);
```
Maybe we should change this to `debug` level? WDYT? I'm not sure if I'd consider it a critical error, and it could flood the logs if the autocomplete interval is running every minute.

mibragimov commented:
> Thanks for taking this one on @mibragimov! Verified with your testing instructions and everything worked as expected.
> 
> I left a couple minor comments in the code. Also, what is our testing strategy in general for the autocomplete request? Are there existing API integration tests for it?

Thanks for the review @alisonelizabeth! I've addressed your feedback with https://github.com/elastic/kibana/pull/140569/commits/f94b90bec283882fb77bbac887a4f469e02ada61. We have existing API integration tests for autocomplete_entities, but they are not very extensive. I've added a few more tests to cover the response format.

mibragimov commented:
Verified the api integration tests through the flaky test runner ~200 times https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/1539

alisonelizabeth commented on test/api_integration/apis/console/autocomplete_entities.ts:
```diff
@@ -13,6 +13,149 @@ import type { FtrProviderContext } from '../../ftr_provider_context';
 
+  const createIndex = async (indexName: string) => {
+    await supertest
+      .post(`/api/console/proxy?method=PUT&path=/${indexName}`)
```
Does it make more sense to use the `es` service for these setup/teardown utils rather than calling the Console API?

mibragimov replied:
Thanks for pointing this out! I updated the tests to use the es service instead of the Console API.

alisonelizabeth commented on test/api_integration/apis/console/autocomplete_entities.ts:
```diff
@@ -129,5 +297,63 @@ export default ({ getService }: FtrProviderContext) => {
+
+      const { body, status } = response;
+      expect(status).to.be(200);
+      expect(body.dataStreams.data_streams.map((ds: any) => ds.name)).to.contain(dataStreamName);
```
Can we avoid using `any` here?

alisonelizabeth commented on test/api_integration/apis/console/autocomplete_entities.ts:
```diff
@@ -129,5 +297,63 @@ export default ({ getService }: FtrProviderContext) => {
+      const { body, status } = response;
+      expect(status).to.be(200);
+      expect(Object.keys(body.legacyTemplates)).to.contain(legacyTemplateName);
+      expect(body.indexTemplates.index_templates.map((it: any) => it.name)).to.contain(
```
Can we avoid using `any` here?

alisonelizabeth commented on test/api_integration/apis/console/autocomplete_entities.ts:
```diff
@@ -129,5 +297,63 @@ export default ({ getService }: FtrProviderContext) => {
+      expect(body.indexTemplates.index_templates.map((it: any) => it.name)).to.contain(
+        indexTemplateName
+      );
+      expect(body.componentTemplates.component_templates.map((ct: any) => ct.name)).to.contain(
```
Can we avoid using `any` here?

alisonelizabeth reviewed (approved):
Thanks for addressing my feedback! I left a couple nits in the API integration tests. Let me know what you think. I'm going to go ahead and approve to unblock you.

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.5|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/145051">](https://github.com/elastic/kibana/pull/145051)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->
